File: src/sarma_cli/engine/audit_slim_graph.py
"""Audit-Slim graph — compact 4-stage pipeline.

recon -> hunter <-> verify -> report. Reuses the subagent node machinery from
the full audit graph so streaming and per-stage event attribution behave
identically.
"""

from __future__ import annotations

from typing import Any, Literal

from langgraph.graph import StateGraph, START, END
from langgraph.types import Command

from sarma_cli.engine.audit_graph import AuditState, _make_subagent_node, _write_audit_event
from sarma_cli.engine.audit_slim_subagents import (
    AUDIT_SLIM_SUBAGENTS,
    AUDIT_SLIM_SUBAGENT_ORDER,
)

DEFAULT_MAX_VERIFY_FEEDBACK = 3


def _verify_router(state: AuditState) -> Command[Literal["hunter", "report"]]:
    """Route weak verify results back to Hunter, or confirmed results to Report."""
    output = (state.get("stage_outputs") or {}).get("verify", "")
    lower = output.lower()
    normalized = lower.lstrip()
    count = state.get("feedback_count", 0)
    needs_hunter = normalized.startswith("needs-hunter") or any(marker in lower for marker in (
        "needs-hunter",
        "needs hunter",
        "not reliable",
        "not confirmed",
        "insufficient",
        "unsupported",
        "false positive",
        "weak",
    ))
    confirmed = normalized.startswith("verified") or any(marker in lower for marker in (
        "verified",
        "reliable finding",
        "ready for reporting",
    ))

    if needs_hunter and not normalized.startswith("verified") and count < DEFAULT_MAX_VERIFY_FEEDBACK:
        next_count = count + 1
        _write_audit_event({
            "type": "audit_route",
            "from": "verify",
            "to": "hunter",
            "loop": "feedback",
            "count": next_count,
        })
        return Command(update={"feedback_count": next_count}, goto="hunter")

    _write_audit_event({
        "type": "audit_route",
        "from": "verify",
        "to": "report",
    })
    return Command(goto="report")


def build_audit_slim_graph(
    model: Any,
    tools: list[Any],
    system_prompt: str = "",
    subagent_specs: list[dict[str, Any]] | None = None,
    subagent_models: dict[str, Any] | None = None,
    subagent_mcp_allow: dict[str, list[str] | None] | None = None,
    subagent_skills: dict[str, object] | None = None,
) -> Any:
    """Build and compile the audit-slim StateGraph.

    Args:
        model: Default LLM for subagents.
        tools: All available MCP tools (each subagent filters by prefix).
        system_prompt: Kept for API parity with build_audit_graph (unused).
        subagent_specs: Slim subagent specs (defaults to AUDIT_SLIM_SUBAGENTS).
        subagent_models: Optional per-agent model overrides.

    Returns:
        Compiled LangGraph usable with astream().
    """
    specs = subagent_specs or AUDIT_SLIM_SUBAGENTS
    spec_map = {s["name"]: s for s in specs}

    builder = StateGraph(AuditState)
    for name in AUDIT_SLIM_SUBAGENT_ORDER:
        node_fn = _make_subagent_node(
            name,
            spec_map[name],
            model,
            tools,
            subagent_models,
            (subagent_mcp_allow or {}).get(name),
            (subagent_skills or {}).get(name),
        )
        builder.add_node(name, node_fn)

    builder.add_node("verify_check", _verify_router)

    # Harness: recon -> hunter <-> verify -> report
    builder.add_edge(START, "recon")
    builder.add_edge("recon", "hunter")
    builder.add_edge("hunter", "verify")
    builder.add_edge("verify", "verify_check")
    builder.add_edge("report", END)

    return builder.compile()

File: src/workflows/auditSlimGraph.ts
/**
 * Audit-Slim graph — compact 4-stage pipeline.
 *
 * recon -> hunter <-> verify -> report. Reuses the subagent node machinery
 * from the full audit graph so streaming and per-stage event attribution
 * behave identically.
 */

import { createAgent } from "langchain";
import type { StructuredToolInterface } from "@langchain/core/tools";
import type { BaseChatModel } from "@langchain/core/language_models/chat_models";
import { StateGraph, START, END, Command } from "@langchain/langgraph";
import {
  AuditState,
  type AuditStateType,
  makeSubagentNode,
  makeRouteAgent,
  routeNext,
  routeNextStructured,
} from "@/workflows/auditGraph";
import {
  AUDIT_SLIM_SUBAGENTS,
  AUDIT_SLIM_SUBAGENT_ORDER,
} from "@/workflows/auditSlimSubagents";
import type { SubagentSpec } from "@/workflows/auditSubagents";
import type { ResolvedSkill } from "@/engine/models";
import { getWriter } from "@langchain/langgraph";
import type { TokenEstimator } from "@/context/tokenizer";

export const DEFAULT_MAX_VERIFY_FEEDBACK = 3;

function writeAuditEvent(data: Record<string, unknown>): void {
  let writer: ((chunk: unknown) => void) | undefined;
  try {
    writer = getWriter();
  } catch {
    return;
  }
  writer?.(data);
}

/** Route weak verify results back to Hunter, or confirmed results to Report. */
function verifyRouterFromDecision(state: AuditStateType, decision: string): Command {
  const output = (state.stage_outputs ?? {}).verify ?? "";
  const lower = output.toLowerCase();
  const normalized = lower.replace(/^\s+/, "");
  const count = state.feedback_count ?? 0;
  const needsHunter =
    normalized.startsWith("needs-hunter") ||
    [
      "needs-hunter",
      "needs hunter",
      "not reliable",
      "not confirmed",
      "insufficient",
      "unsupported",
      "false positive",
      "weak",
    ].some((marker) => lower.includes(marker));

  if (
    (decision === "hunter" ||
      (!decision && needsHunter && !normalized.startsWith("verified"))) &&
    count < DEFAULT_MAX_VERIFY_FEEDBACK
  ) {
    const nextCount = count + 1;
    writeAuditEvent({
      type: "audit_route",
      from: "verify",
      to: "hunter",
      loop: "feedback",
      count: nextCount,
    });
    return new Command({ update: { feedback_count: nextCount }, goto: "hunter" });
  }

  writeAuditEvent({ type: "audit_route", from: "verify", to: "report" });
  return new Command({ goto: "report" });
}

export interface BuildAuditSlimGraphOptions {
  systemPrompt?: string;
  subagentSpecs?: SubagentSpec[];
  subagentModels?: Record<string, BaseChatModel> | null;
  subagentMcpAllow?: Record<string, string[] | null> | null;
  subagentSkills?: Record<string, ResolvedSkill | null> | null;
  structuredRouting?: boolean;
  maxPriorStageTokens?: number;
  estimateText?: TokenEstimator;
  compileKwargs?: Record<string, unknown>;
}

/** Build and compile the audit-slim StateGraph. */
export function buildAuditSlimGraph(
  model: BaseChatModel,
  tools: StructuredToolInterface[],
  options: BuildAuditSlimGraphOptions = {},
) {
  const specs = options.subagentSpecs ?? AUDIT_SLIM_SUBAGENTS;
  const specMap = new Map(specs.map((s) => [s.name, s]));
  const structuredRouting = options.structuredRouting ?? true;

  const verifyRouteAgent = structuredRouting
    ? makeRouteAgent(model, "audit_slim_verify_router")
    : null;

  type NodeFn = (state: AuditStateType, config: never) => Promise<Partial<AuditStateType>>;
  const builder = new StateGraph(AuditState);
  const g = builder as unknown as {
    addNode: (name: string, fn: NodeFn, opts?: { ends?: string[] }) => void;
    addEdge: (from: string, to: string) => void;
    compile: (opts?: Record<string, unknown>) => ReturnType<StateGraph<typeof AuditState.spec>["compile"]>;
  };

  for (const name of AUDIT_SLIM_SUBAGENT_ORDER) {
    const nodeFn = makeSubagentNode(name, specMap.get(name)!, model, tools, {
      subagentModels: options.subagentModels,
      allowedMcpServers: (options.subagentMcpAllow ?? {})[name],
      skill: (options.subagentSkills ?? {})[name],
      maxPriorStageTokens: options.maxPriorStageTokens,
      estimateText: options.estimateText,
    }) as unknown as NodeFn;
    g.addNode(name, nodeFn);
  }

  const verifyCheck: NodeFn = async (state) => {
    writeAuditEvent({
      type: "subagent_start",
      name: "verify_check",
      description: "same-model structured router: hunter | report",
    });
    const output = (state.stage_outputs ?? {}).verify ?? "";
    let decision: string;
    if (verifyRouteAgent === null) {
      decision = routeNext(output, new Set(["hunter", "report"]));
    } else {
      try {
        decision = await routeNextStructured(
          verifyRouteAgent,
          "verify",
          output,
          new Set(["hunter", "report"]),
        );
      } catch {
        decision = routeNext(output, new Set(["hunter", "report"]));
      }
    }
    const next = verifyRouterFromDecision(state, decision) as unknown as Partial<AuditStateType>;
    writeAuditEvent({ type: "subagent_complete", name: "verify_check" });
    return next;
  };

  g.addNode("verify_check", verifyCheck, { ends: ["hunter", "report"] });

  // Harness: recon -> hunter <-> verify -> report
  g.addEdge(START, "recon");
  g.addEdge("recon", "hunter");
  g.addEdge("hunter", "verify");
  g.addEdge("verify", "verify_check");
  g.addEdge("report", END);

  return g.compile(options.compileKwargs);
}
